control-plane/agents/core/src/core/reconciler/pool/mod.rs

```diff
@@ -0,0 +1,112 @@
+use crate::core::{
+    specs::{OperationSequenceGuard, SpecOperations},
+    task_poller::{PollContext, PollPeriods, PollResult, PollTimer, PollerState, TaskPoller},
+    wrapper::ClientOps,
+};
+use common_lib::types::v0::{
+    message_bus::{CreatePool, NodeStatus},
+    store::{pool::PoolSpec, OperationMode, TraceSpan},
+};
+use parking_lot::Mutex;
+use std::sync::Arc;
+
+/// Pool Reconciler loop which:
+/// 1. recreates pools which are not present following a mayastor restart
+#[derive(Debug)]
+pub struct PoolReconciler {
+    counter: PollTimer,
+}
+impl PoolReconciler {
+    /// Return new `Self` with the provided period
+    pub fn from(period: PollPeriods) -> Self {
+        PoolReconciler {
+            counter: PollTimer::from(period),
+        }
+    }
+    /// Return new `Self` with the default period
+    pub fn new() -> Self {
+        Self::from(1)
+    }
+}
+
+#[async_trait::async_trait]
+impl TaskPoller for PoolReconciler {
+    async fn poll(&mut self, context: &PollContext) -> PollResult {
+        let mut results = vec![];
+        for pool in context.registry().specs().get_locked_pools() {
+            if pool.lock().status().created() {
+                results.push(missing_pool_state_reconciler(pool, context).await)
+            }
+        }
+        Self::squash_results(results)
+    }
+
+    async fn poll_timer(&mut self, _context: &PollContext) -> bool {
+        self.counter.poll()
+    }
+}
+
+/// If a pool has a spec but not state, it means that the mayastor instance where the pool should
+/// exist does not have the pool open.
+/// This can happen if the pool is destroyed under the control plane, or if mayastor
+/// crashed/restarted.
+/// In such a case, we issue a new create pool request against the mayastor instance where the pool
+/// should exist.
+#[tracing::instrument(skip(pool_spec, context), fields(pool.uuid = %pool_spec.lock().id))]
+async fn missing_pool_state_reconciler(
+    pool_spec: Arc<Mutex<PoolSpec>>,
+    context: &PollContext,
+) -> PollResult {
+    if !pool_spec.lock().status().created() {
+        // nothing to do here
+        return PollResult::Ok(PollerState::Idle);
+    }
```

Comment:
Is it possible for the pool spec to be stuck in "Creating"? For example if Mayastor crashed/restarted during pool creation. If so, should we attempt to create it?

Reply:
I don't think that can happen anymore following your "delete spec when creation fails" changes. But if it did I guess the operator would also retry the request and it would eventually succeed.

Reply:
Yeah I thought so - just wanted to double check.